### PR TITLE
rpc: remove the eth_createAccessList authorization-count guard

### PR DESCRIPTION
**Pull request by Sahil-4555**
`eth_createAccessList` bounded the authorization count by dividing the available gas by 25000. Once Amsterdam is active an authorization really costs 7816, so the check refused calls that had enough gas - including the figure `eth_estimateGas` returns for the same call.

The count check is gone. The request's intrinsic gas is now checked against its effective gas limit, before any authority is recovered.

### Why not just fix the divisor

Recovering an authority is an ECDSA operation, and the old code ran one per entry before any gas accounting, so a list that could never execute still paid for every recovery.

Repricing the divisor would not have fixed that, and there is no honest value for it either. To avoid false rejections it has to be 7816, the floor — but that floor only describes a list naming one already-delegated authority. A real delegation to a fresh account costs about 235,600, because [EIP-2780](https://eips.ethereum.org/EIPS/eip-2780#eip-7702-authorization-processing) adds `ACCOUNT_WRITE`, new-account and delegation-byte charges [at most once per authority](https://github.com/erigontech/erigon/blob/4c6a63e359b360c81b9026edfba3f154c7db8a7d/execution/protocol/txn_executor.go#L938-L960). So a correct divisor admits roughly thirty times what can actually run.

The check that replaces it takes the per-authorization cost [from the block's fork rules](https://github.com/erigontech/erigon/blob/4c6a63e359b360c81b9026edfba3f154c7db8a7d/execution/protocol/mdgas/intrinsic_gas.go#L266-L276), and `ToMessage` supplies the effective gas limit, so a request that sends no `gas` field is still bounded by the RPC gas cap. It leaves the access list out: the tracer strips excluded addresses from that list before the executor prices it, so the check stays at or below what the executor charges. It reports what the executor reports:

```
intrinsic gas too low: have 21000, want 85344
```

It also fixes a smaller bug on the way: with `--rpc.gascap=0` the old guard divided zero by 25000 and refused every authorization list outright.

The spec asks for none of this: [`eth_createAccessList`](https://github.com/ethereum/execution-apis/blob/465d1b98d43e94ff3d57e904fd7d4bea7f6b804c/src/eth/execute.yaml#L68-L95) puts no limit on `authorizationList` length relative to `gas` and defines no matching error.

### Cost

`-count=10 -cpu=1`, medians, 65,536 authorizations:

| | ns/op | per entry | allocs |
|---|---:|---:|---:|
| check before recovery | 3,754,578 | 57 ns | 52 |
| check after recovery | 2,795,214,094 | 42,652 ns | 196,720 |

### Testing

`TestExcludeAuthoritiesOrder` asserts an empty exclusion set for an explicit gas limit below the intrinsic cost and for omitted gas under a finite RPC gas cap, with an affordable request beside them that recovers eight distinct authorities into the set, so an assertion that never looked at the set could not pass the other two. Moving the check after the loop fails both refused cases, since the set then holds an authority.

`TestCreateAccessListAuthGas` pins that an `eth_estimateGas` result is accepted on Amsterdam and pre-Amsterdam, and that an under-gassed request is refused by the intrinsic check, with and without a `gas` field. `TestCreateAccessListExcludedAccessList` pins that an access list entry the tracer drops, the sender with 200 storage keys, is not charged against the gas limit. `BenchmarkCreateAccessListAuthGas` is the table above.

### CODE REVIEW

File: rpc/jsonrpc/eth_call.go [hunk 1]
```diff
@@ -1010,13 +1010,6 @@ func (api *APIImpl) CreateAccessList(ctx context.Context, args ethapi2.CallArgs,
...
 	// EIP-7702: authority addresses are pre-warmed in state transition, so exclude them from the access list
 	if len(args.AuthorizationList) > 0 {
-		gasCap := api.GasCap
-		if args.Gas != nil && uint64(*args.Gas) < gasCap {
-			gasCap = uint64(*args.Gas)
-		}
-		if uint64(len(args.AuthorizationList)) > gasCap/params.CallNewAccountGas {
-			return nil, errors.New("insufficient gas to process all authorizations")
-		}
 		rules := blockCtx.Rules(chainConfig)
```
Comment (yperbasis) [inline rpc/jsonrpc/eth_call.go:1013]: **Check intrinsic gas before entering the authority recovery loop.**

After this guard is removed, every authorization reaches `RecoverSigner()` before `ApplyMessage` checks intrinsic gas. The recovery loop also does not check context cancellation. The later gas error cannot limit the cryptographic work already done.

I reproduced this locally with 100,000 copies of a valid authorization, using the default 50M RPC gas cap. The JSON-encoded call arguments occupy 24.9 MB, below the 32 MB HTTP limit. Timings below measure `CreateAccessList` itself:

| Request | This PR | Original guard restored |
|---|---:|---:|
| `gas = 21000` | 2.46 s | 48 µs |
| `gas` omitted | 2.44 s | 30 µs |

All four calls reject the request. Omitting `gas` still left a useful bound in the old code: the default cap limited the recovery loop to 2,000 authorizations.

Please run fork-aware intrinsic-gas validation before this loop, then recover authorities after it passes. Nethermind uses this order: [`ValidateStatic` runs first](https://github.com/NethermindEth/nethermind/blob/0d43dbdf8987ff4c55f49e97cee129071e0b8258/src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs#L200-L263), and [authority recovery happens later in `ProcessDelegations`](https://github.com/NethermindEth/nethermind/blob/0d43dbdf8987ff4c55f49e97cee129071e0b8258/src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs#L746-L750).

Please add regression coverage for rejecting oversized lists before recovery, both with an explicit low gas limit and with `gas` omitted under a finite RPC gas cap.

**Review by yperbasis — CHANGES_REQUESTED** [review]
Please validate intrinsic gas against the effective gas limit, including the RPC gas cap, before recovering authorities. Use the active fork rules so valid Amsterdam gas estimates remain accepted.

The existing `TestCreateAccessList*` tests pass, but they do not cover the cost of recovering authorities before an oversized request is rejected.

File: rpc/jsonrpc/eth_call.go (hunk 1 above)
Comment (Sahil-4555) [reply rpc/jsonrpc/eth_call.go]: Done - intrinsic gas is now validated before the recovery loop, and the loop only runs once it passes.

`ToMessage` already resolves the effective gas limit, so a request that sends no `gas` field is bounded by the RPC gas cap, and `mdgas.IntrinsicGas` prices the authorizations under the block's fork rules, so an Amsterdam estimate is still accepted. The check itself is the same comparison `preCheck` makes, `max(ExecutionGas, FloorGasCost)` against the limit, and it reports the same error, so an under-gassed request now gets `intrinsic gas too low: have 21000, want 85344` instead of a count it never chose.

I looked for an existing entry point to call rather than adding one. There is none exposed: `calcIntrinsicGas` is an unexported method on `TxnExecutor`, and every other caller (txpool, shutter, the AA path) builds `mdgas.IntrinsicGasCalcArgs` itself. So the new helper does the same, from the message rather than the raw args, which keeps the gas-cap and calldata handling in `ToMessage` instead of copying it.

Regression coverage is `TestCreateAccessListAuthorizationsAffordable`, with both cases you asked for: an explicit low gas limit, and `gas` omitted under a finite RPC gas cap. A third case pins the cost itself — 65536 authorizations must be rejected in under a second, which fails at ~2.5s without the check and passes at ~10ms with it.

Numbers here match yours closely: 2.4s to 2.7s before, tens of microseconds after.

File: rpc/jsonrpc/eth_call_test.go [hunk 2]
```diff
@@ -1758,6 +1758,151 @@ func TestCallRejectsBlobHashesBeforeCancun(t *testing.T) {
...
+
+		start := time.Now()
+		_, err := api.CreateAccessList(context.Background(), args, nil, nil, nil)
+		elapsed := time.Since(start)
+
+		require.ErrorIs(t, err, protocol.ErrIntrinsicGas)
+		// Recovering this many authorities runs into seconds, while the check that
+		// replaces them is arithmetic. The bound is loose enough to survive a slow
+		// machine and still tell the two apart.
+		require.Less(t, elapsed, time.Second, "authorities were recovered before the request was rejected")
```
[the remark this answers is not in the document]
Comment (Sahil-4555) [reply rpc/jsonrpc/eth_call_test.go]: Agreed on both counts, and the first one is measurable: `ToMessage` allocates the authorization slice and calls `ToAuthorization` on every entry, so the decode is linear in the list length and sits inside the timed window. The early-rejection path was never O(1). Under `-race` it measured 0.04s against the 1s bound, so there was margin, but the baseline grows with the input exactly as you describe.

The second direction is the weaker one and the reason I dropped the assertion: the regression it was meant to catch ran 2.4s here, only about 2.5x above the bound, so quicker hardware could have run all 65536 recoveries inside a second and let it pass silently.

Replaced it with a benchmark, per your suggestion. Measured both ways at 65536 entries:

| | ns/op | per entry |
|---|---:|---:|
| check before the loop | 3,640,711 | ~55 ns |
| check removed | 2,591,065,499 | ~39.5 us |

A check that moves back after the recovery loop takes the per-entry cost from tens of nanoseconds to tens of microseconds, which is far louder than a threshold and has nothing to trip over.

On the injectable recovery helper: I left that out. It means a mutable package-level seam in the RPC path existing only for a test, and there is no precedent for that pattern in this tree. The two cases the reviewer asked for still assert deterministically on the error, so the behaviour stays covered without it. Happy to add the seam if a hard ordering assertion is wanted.

[the remark this answers is not in the document]
Comment (Sahil-4555) [reply rpc/jsonrpc/eth_call_test.go]: Measured with `-count=10 -cpu=1`, medians:

| entries | check before the loop | check removed | factor |
|---:|---:|---:|---:|
| 1,024 | 74,831 ns | 38,389,708 ns | 513x |
| 16,384 | 1,431,014 ns | 646,546,400 ns | 452x |
| 65,536 | 3,754,578 ns | 2,795,214,094 ns | 744x |

Per entry that is 57-87 ns against 37-43 us, and allocations at 65,536 go from 52 to 196,720. The per-entry cost is what the ECDSA recovery adds, so a check moved back after the loop is three orders of magnitude in the numbers.

**Review by yperbasis — CHANGES_REQUESTED** [review]
Intrinsic gas is now checked before authority recovery. The benchmark and repeated measurements support the performance fix, and removing the wall-clock assertion addresses its reliability problem.

The remaining request in [the review thread](https://github.com/erigontech/erigon/pull/23985#discussion_r4015136701) is deterministic coverage of recovery ordering. The current tests only check that the request is eventually rejected.

I checked this on `97c8c8f` by moving `checkIntrinsicGas` after the authority recovery loop in a temporary Go overlay. All `TestCreateAccessList*` tests and `BenchmarkCreateAccessListAuthGas` still passed. The 65,536-entry benchmark took 1.54 seconds with that mutation. The benchmark reports the slowdown, but its error assertion does not fail when the expensive recovery work returns.

Please add a deterministic test asserting zero recovery calls for both an explicit low gas limit and omitted gas under a finite RPC gas cap. Include an affordable request that reaches recovery to verify the test setup. A recovery function passed to a small private helper is one option; this does not require a mutable package-level variable.

Verify that the new test fails when validation moves after recovery and passes with the current ordering.

File: rpc/jsonrpc/eth_call_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (Sahil-4555) [reply rpc/jsonrpc/eth_call_test.go]: Added in [`1bb9451`](https://github.com/erigontech/erigon/pull/23985/commits/1bb94510c891ea72771e936db22a1bf139193de6). Your framing solved my objection: the recovery function is a parameter on a private helper, so there is no mutable package-level state.

The check and the loop moved into `excludeAuthorities(msg, chainRules, excl, recover)`. Production passes `(*types.Authorization).RecoverSigner`; the test passes a counting stub.

`TestExcludeAuthoritiesOrder` asserts zero recoveries for both cases you named, an explicit gas limit under the intrinsic cost and no gas field under a finite RPC gas cap, and a third affordable case recovers all 8 and populates the exclusion set, so a counter that never moved could not make the other two pass.

Verified against your mutation. With the check after the loop both cases fail on the count, 64 and 4096 respectively, naming exactly how many recoveries were paid for. With the current ordering all three pass in 0.00s.

One incidental simplification: the loop reads the authorizations off the message instead of calling `ToAuthorization` per entry. `ToMessage` has already converted them by that point, and it errors on a malformed entry rather than skipping it, so the per-entry conversion and its `continue` were unreachable.

**Review by yperbasis — APPROVED** [review]
Re-reviewed `1bb9451`. The recovery-order coverage requested in [my previous review](https://github.com/erigontech/erigon/pull/23985#pullrequestreview-5209696097) is now in place.

`excludeAuthorities` checks intrinsic gas before invoking the recovery function. Passing that function as a parameter avoids mutable package-level state. The new test covers both rejected requests and an affordable request that performs all 8 recoveries and populates the exclusion set.

Verified locally:

- `go test ./rpc/jsonrpc -run '^Test(ExcludeAuthoritiesOrder|CreateAccessList)' -bench '^BenchmarkCreateAccessListAuthGas$' -benchtime=1x -count=1` passes.
- Moving validation after recovery in a temporary Go overlay makes `TestExcludeAuthoritiesOrder` fail on both rejection cases: 64 and 4096 recoveries instead of zero. The affordable case still passes.

No remaining blocking findings from this review. CI is still running.

File: rpc/jsonrpc/eth_call.go [hunk 3]
```diff
@@ -898,6 +900,69 @@ type accessListResult struct {
...
+// beside it is gated on CheckGas, which ToMessage leaves false, so enforcing it
+// here would reject calls the executor would run.
+func checkIntrinsicGas(msg *types.Message, chainRules *chain.Rules) error {
+	contractCreation := msg.To().IsNil()
+	accessList := msg.AccessList()
+	intrinsic, overflow := mdgas.IntrinsicGas(mdgas.IntrinsicGasCalcArgs{
+		Data:               msg.Data(),
+		AuthorizationsLen:  uint64(len(msg.Authorizations())),
+		AccessListLen:      uint64(len(accessList)),
+		StorageKeysLen:     uint64(accessList.StorageKeys()),
```
Comment (yperbasis) [inline rpc/jsonrpc/eth_call.go]: The tracer strips excluded addresses from this list before the executor prices it, so these two fields make the check stricter than the executor. Dropping them, together with the `accessList` variable, keeps the check a lower bound; the authorization term alone bounds the recoveries.

**Review by yperbasis — CHANGES_REQUESTED** [review]
The intrinsic-gas check prices the caller's raw access list, but `NewAccessListTracer` drops every entry whose address is in `excl` (sender, precompiles, authorities), so the executor prices a smaller list. The check can therefore reject a request that `main` serves.

Repro on Amsterdam rules: one signed authorization, and an access list with a single entry, the sender with 200 storage keys.

| gas | main | this PR |
|---|---|---|
| omitted | ok, gasUsed 250606 | ok, gasUsed 250606 |
| 280000 | ok, gasUsed 250606 | `intrinsic gas too low: have 280000, want 836596` |
| 450000 | ok | same rejection |

An `eth_estimateGas` figure still passes because the estimator prices the raw list too, so this needs a hand-written list, but the docstring's "the same comparison the executor makes" does not hold.

Please:
- drop `AccessListLen` and `StorageKeysLen` from `checkIntrinsicGas` (see the inline comment). The check then never exceeds what the executor charges, and recoveries stay bounded by `gas / perAuthCost`, which is all the guard needs. Verified locally: the 280000 case is accepted again, and all `TestCreateAccessList*` tests and `TestExcludeAuthoritiesOrder` still pass;
- add a regression test with the sender in the access list and a gas limit between the two figures, expecting success;
- adjust the docstring sentence accordingly.

Nits, non-blocking:
- `97c8c8f` moved the benchmark into `eth_call_test.go`. CLAUDE.md prefers a separate `_bench_test.go` file, and `witness_cache_bench_test.go` is the local precedent. The first version had it right.
- The `recover` parameter shadows the builtin; `recoverSigner` reads better.
- The `recoverAuthority` comment and the two-paragraph `checkIntrinsicGas` docstring can shrink; the EIP-7825 sentence is the one worth keeping.
- `TestCreateAccessListAuthorizationsAffordable` repeats `TestExcludeAuthoritiesOrder` and the under-gassed subtest; consider folding it. It also has a stray blank line before its closing brace.
- Worth a line in the description: with `--rpc.gascap=0` the old guard divided zero by 25000 and rejected every authorization list; this PR fixes that.
- Alex's inline-the-helper comment is still unanswered. The restructure supersedes it, but a short reply would close it.

File: rpc/jsonrpc/eth_call.go (hunk 3 above)
Comment (Sahil-4555) [reply rpc/jsonrpc/eth_call.go]: You are right, this was a real bug. Dropped both fields and the `accessList` variable in https://github.com/erigontech/erigon/pull/23985/commits/5a0694e19f10c7034632e066077818f42eea7d7a.

Reproduced your case before changing anything and got the same number, `intrinsic gas too low: have 280000, want 836596`, and confirmed the request goes through once the check is removed. `NewAccessListTracer` skips any address in `excl`, so the sender entry never reached the executor while my check was still charging 200 storage keys for it.

Added `TestCreateAccessListExcludedAccessList` with the sender in the access list and gas 280000, expecting success. Recoveries are still bounded by the authorization term, and `TestExcludeAuthoritiesOrder` still fails at 64 and 4096 if the check moves after the loop. Docstring now says the access list is left out and why.

**Comment by Sahil-4555** [comment]
Thanks for the careful review. All of it is in [`5a0694e`](https://github.com/erigontech/erigon/pull/23985/commits/5a0694e19f10c7034632e066077818f42eea7d7a) except one, notes below.

Done:

- `recover` renamed to `recoverSigner`. Shadowing the builtin was sloppy, agreed.
- Both comments shrunk. Kept the EIP-7825 sentence as you suggested.
- `TestCreateAccessListAuthorizationsAffordable` folded away with its stray blank line. Its low-gas case only repeated the under-gassed subtest, so that one is gone. The no-gas-field case moved into `TestCreateAccessListAuthGas` rather than being dropped, since it is the only place that shows `api.GasCap` actually reaching `ToMessage`; the helper test cannot show the wiring.
- Added the `--rpc.gascap=0` line to the description. Good catch, `0/25000` is `0` so every list was refused.
- Replied on Alex's thread.

Not done: the benchmark is still in `eth_call_test.go`. It was moved there deliberately to sit next to the cases it backs, but you are right that CLAUDE.md says otherwise and `witness_cache_bench_test.go` is the precedent. Happy to move it back, just say so.

**Comment by AskAlexSharov** [comment]
No correctness bug found. The replacement check is at or below what the executor charges on every argument, so it cannot refuse a call the executor would run. Two points.

**The `recoverAuthority` parameter is a test seam production code does not need, and the test already proves the ordering without it.** `require.Empty(t, excl)` sits one line under `require.Zero(t, calls)`, and it is the assertion that does the work: with the counter check neutered and `checkIntrinsicGas` moved after the loop, both refusal cases still fail on `Should be empty, but was map[0x8AcF...]`. Drop the named type, the parameter and its comment, and call `auth.RecoverSigner()` directly. The one thing the counter says that `excl` cannot is in the affordable case — `require.Equal(t, 8, calls)` against a single repeated authority. Sign eight distinct keys there and `require.Len(t, excl, 8)` says it better.

**The benchmark ended up in `eth_call_test.go`.** The repo asks for benchmarks in their own `xyz_bench_test.go`, and this package already has `witness_cache_bench_test.go`. The branch created `eth_call_bench_test.go` and then moved the benchmark out of it, so this is a `git mv` back.

**Comment by Sahil-4555** [comment]
The red check is `race-tests / tests-linux (ubuntu-latest, consensus, parallel)`. It fails on `TestExecutionPayloadIndexWriteHasSingleNotificationOwner` in `cl/phase1/forkchoice` with `expected: 1, actual: 2`, which is the flake #24019 is already deflaking.

Not from this PR. It touches three files under `rpc/jsonrpc` and nothing else, and no code in `cl/` calls `CreateAccessList`, `excludeAuthorities` or `checkIntrinsicGas`, so the changed code is not reachable from that test. The branch is also level with main, so the forkchoice code here is the same as main's.

Checked before concluding that: ran the full package with `-race` five times on this branch and on a clean main worktree, green every time, and it only shows up in the parallel consensus job.

Will re-run the job once #24019 is merged.

**Review by yperbasis — APPROVED** [review]
